sycl/doc/extensions/proposed/sycl_ext_intel_grf_size.asciidoc

```diff
@@ -0,0 +1,178 @@
+Attempting to use this extension in kernels that run on other devices or
+backends may result in undefined behavior.  Be aware that the compiler
+is not able to issue a diagnostic to warn you if this happens.
+
+== Overview
+
+There are devices where the size of the general register file (GRF) used by a kernel is 
+configurable. Developers may want to change the GRF size based on their
+application. This extension adds the kernel property `grf_size` which provides a way
+to specify the GRF size for a SYCL kernel, and the kernel property `grf_size_automatic`
+which provides a way to request a valid GRF size be automatically chosen.
+
+The properties described in this extension are advanced features that most applications
+should not need to use. In most cases, applications get the best performance
+without using these properties.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_INTEL_GRF_SIZE` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Properties
+
+|===
+|Property|Description
+
+|`grf_size`
+|The `grf_size` kernel property provides a way to specify the GRF size used by a kernel. 
+It accepts a single unsigned integer value.
+
+|`grf_size_automatic`
+| The `grf_size_automatic` kernel property adds the requirement that the kernel use any of the supported GRF sizes. The manner in which the GRF size is selected is implementation-defined.
+
+|===
+
+At most one of the `grf_size` and `grf_size_automatic` properties may be associated with a kernel.
+
+If a kernel is not associated with a `grf_size` or `grf_size_automatic` property, the manner in which the GRF size is selected is implementation-defined.
+
+The properties are defined as follows:
+```c++
+namespace sycl::ext::intel::experimental {
+
+struct grf_size_key {
+  template <unsigned int Size>
+  using value_t = 
+      oneapi::experimental::property_value<grf_size_key, 
+                                           std::integral_constant<unsigned int, Size>>;
+};
+
+struct grf_size_automatic_key {
+  using value_t = 
+      oneapi::experimental::property_value<grf_size_automatic_key>;
+};
+
+template <unsigned int Size>
+inline constexpr grf_size_key::value_t<Size> grf_size;
+
+inline constexpr grf_size_automatic_key::value_t grf_size_automatic;
+
+} // namespace sycl::ext::intel::experimental
+```
+The supported values are as follows:
+[%header,cols="1,5"]
+|===
+|GPU |Supported Values
```

Comment:
GPU thread (which has register resource) is used to schedule each sub-group. And after sub-group size is determined, the kernel (work item in the subgroup) should use all the register as much as possible, so while do we need a setting to limit the using of registers?  do you have example to show the usage of this new property. Thanks.

Reply:
the problem is the GPU driver decides how many registers are available to the thread, there are two options, small (256) and large (512) and right now it always picks small. large may help if there is high register pressure. i don't have a specific example i can share here, but we have many cases internally showing improved performance with large mode

Reply:
In https://www.intel.com/content/www/us/en/docs/oneapi/optimization-guide-gpu/2023-1/small-register-mode-vs-large-register-mode.html,  " set_kernel_properties(kernel_properties::use_large_grf); " is used to guide compiler to use large register file for specific kernel,  so why do we need anther property to set small(128) or large(256)? Thanks.

Reply:
I'm moving this conversation thread offline.

Reply:
@gmlueck Are you okay to merge this PR, or should we wait for the outcome of the offline conversation? Thanks

Reply:
I think we can merge this PR.

The resolution of the offline discussion may affect other things, but not this PR.

Reply:
Sounds good @intel/llvm-gatekeepers can we merge this one? Thanks!